pkg/upgradehelper/vmlivemigratedetector/detector.go

```diff
@@ -119,3 +154,115 @@ func (d *VMLiveMigrateDetector) Run(ctx context.Context) error {
 
 	return nil
 }
+
+func (d *VMLiveMigrateDetector) getAllNodes(ctx context.Context) ([]*v1.Node, error) {
+	// filter out witness nodes as they are not available for vm migration
+	nodeReq, err := labels.NewRequirement(harvNodeController.HarvesterWitnessNodeLabelKey, selection.DoesNotExist, nil)
+	if err != nil {
+		return nil, err
+	}
+	nodeList, err := d.virtClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: nodeReq.String()})
+	if err != nil {
+		return nil, err
+	}
+	nodes := make([]*v1.Node, 0, nodeList.Size())
+	for i := range nodeList.Items {
+		nodes = append(nodes, &nodeList.Items[i])
+	}
+	return nodes, nil
+}
+
+func (d *VMLiveMigrateDetector) getVMIs(ctx context.Context, opts metav1.ListOptions) ([]*kubevirtv1.VirtualMachineInstance, error) {
+	vmiList, err := d.virtClient.VirtualMachineInstance("").List(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+	vmis := make([]*kubevirtv1.VirtualMachineInstance, 0, len(vmiList.Items))
+	for i := range vmiList.Items {
+		vmis = append(vmis, &vmiList.Items[i])
+	}
+	return vmis, nil
+}
+
+func (d *VMLiveMigrateDetector) getVMIListOptions(nodes []*v1.Node) (metav1.ListOptions, error) {
+	// All VMs are non-migratable if there is only one node
+	if len(nodes) == 1 {
+		return metav1.ListOptions{}, nil
+	}
+	nodeReq, err := labels.NewRequirement("kubevirt.io/nodeName", selection.Equals, []string{d.nodeName})
+	if err != nil {
+		return metav1.ListOptions{}, err
+	}
+	notUpgradeReq, err := labels.NewRequirement("harvesterhci.io/upgrade", selection.DoesNotExist, nil)
+	if err != nil {
+		return metav1.ListOptions{}, err
+	}
+	return metav1.ListOptions{
+		LabelSelector: labels.NewSelector().Add(*nodeReq).Add(*notUpgradeReq).String(),
+	}, nil
+}
+
+func getNonLiveMigratableVMNames(vmis []*kubevirtv1.VirtualMachineInstance, nodes []*v1.Node) ([]string, error) {
+	// All VMs are non-migratable if there is only one node
+	if len(nodes) == 1 {
+		names := make([]string, 0, len(vmis))
+		for _, vmi := range vmis {
+			names = append(names, vmi.Namespace+"/"+vmi.Name)
+		}
+		return names, nil
+	}
+	return virtualmachineinstance.GetAllNonLiveMigratableVMINames(vmis, nodes)
+}
+
+func getRestoreVMNames(vmis []*kubevirtv1.VirtualMachineInstance, nonMigratableVMNames []string) []string {
+	restoreVMs := make([]string, 0, 0)
+
+	// filter out paused VMs and upgrade repo VM from the non-migratable VMs
+	excludeVMs := make(map[string]struct{})
+	for _, vmi := range vmis {
+		if vmPaused.IsTrue(vmi) || vmi.Labels["harvesterhci.io/upgrade"] != "" {
+			namespacedName := fmt.Sprintf("%s/%s", vmi.Namespace, vmi.Name)
+			excludeVMs[namespacedName] = struct{}{}
+		}
+	}
+	for _, name := range nonMigratableVMNames {
+		if _, exist := excludeVMs[name]; !exist {
+			restoreVMs = append(restoreVMs, name)
+		}
+	}
+	return restoreVMs
+}
+
+func (d *VMLiveMigrateDetector) patchUpgradeCR(ctx context.Context, restoreVMNames []string, nodeUID string) error {
+	if len(restoreVMNames) == 0 {
```

Comment:
In this case, the patch might be still needed, if the ugprade CR contains a related annotation which was patched before.

is here a way like ` if !DeepEqual  { client.Update } ` ?